map_view/display-device-location-with-nmea-data-sources/src/main/java/com/esri/samples/display_device_location_with_nmea_data_sources/DisplayDeviceLocationWithNmeaDataSourcesController.java

```diff
@@ -0,0 +1,223 @@
+              timeline = new Timeline();
+              timeline.setCycleCount(-1); // loop count
+              // push the mock data NMEA sentences into the data source every 250 ms
+              timeline.getKeyFrames().add(new KeyFrame(Duration.millis(250), event -> {
+
+                // note: you can also use real-time NMEA sentences obtained via a GPS dongle
+                nmeaLocationDataSource.pushData(nmeaSentences.get(count++).getBytes(StandardCharsets.UTF_8)); // post increment step
+                // reset the count after the last data point is reached
+                if (count == nmeaSentences.size()) {
+                  count = 0;
+                }
+
+              }));
+
+              // start the timeline
+              timeline.play();
+              // handle UI interactions
+              startButton.setDisable(true);
+              stopButton.setDisable(false);
+
+              nmeaLocationDataSource.removeStatusChangedListener(this);
+
+            }
+          }
+        };
+
+        // initialize the location data source and prepare to begin receiving location updates when data is pushed. As
+        // updates are received, they will be displayed on the map
+        nmeaLocationDataSource.addStatusChangedListener(listener);
+        nmeaLocationDataSource.startAsync();
+
+      } catch (Exception e) {
+        new Alert(Alert.AlertType.ERROR, e.getCause().getMessage()).show();
+        e.printStackTrace();
+      }
+
+    } else {
+      new Alert(Alert.AlertType.ERROR, "File not found").show();
+    }
+
+  }
+
+  /**
+   * Obtains NMEA satellite information from the NMEA location data source, and displays satellite information on the app.
+   */
+  private void setupSatelliteChangedListener() {
+
+    HashSet<Integer> uniqueSatelliteIds = new HashSet<>();
+
+    nmeaLocationDataSource.addSatellitesChangedListener(satellitesChangedEvent -> {
+
+      // get satellite information from the NMEA location data source every time the satellites change
+      nmeaSatelliteInfo = satellitesChangedEvent.getSatelliteInfos();
+      // set the text of the satellite count label
+      satelliteCount.setText("Satellite count: " + nmeaSatelliteInfo.size());
+
+      for (NmeaSatelliteInfo satInfo : nmeaSatelliteInfo) {
+        // collect unique satellite ids
+        uniqueSatelliteIds.add(satInfo.getId());
+        // sort the ids numerically
+        List<Integer> sortedIds = new ArrayList<>(uniqueSatelliteIds);
+        Collections.sort(sortedIds);
+        // display the satellite system and id information
+        systemInfo.setText("System: " + satInfo.getSystem());
+        satelliteID.setText("Satellite IDs: " + sortedIds);
+      }
+    });
+  }
+
+  /**
+   * Stops displaying the mock data location, stops receiving location data, and stops the timeline.
+   */
+  @FXML
+  private void stop() {
+
+    // stop receiving and displaying location data
+    nmeaLocationDataSource.stop();
```

Comment:
I think it might be safer to stop the timeline then the location data source otherwise your timeline might have an event and try to push a message to the now stopped nmea location data source.